Analysis pipelines for the pilot phase (V0) of Cattle Cell Atlas (CattleCA) 

1. Introduction

As part of the Farm animal Genotype-Tissue Expression (FarmGTEx) project, we describe a comprehensive reference map comprising 1,793,854 cells from 59 cattle tissues, spanning both sexes and multiple developmental stages. This map, generated by single-cell/nucleus RNA sequencing, identifies 131 distinct cell types, revealing intra- and inter-tissue cellular heterogeneity in gene expression, transcription factor regulation, and intercellular communication. Integrative analysis with genetic variants that underpin various bovine monogenic and complex traits uncovers cell types of relevance, such as spermatocytes responsible for sperm motilities and excitatory neurons for milk fat yield. Comparative analysis reveals similarities in gene expression between cattle and humans at single-cell resolution, allowing for detection of relevant cell types for studying human complex phenotypes. This cattle cell atlas will serve as a key resource for cattle genetics and genomics, immunology, comparative biology, and ultimately human biomedicine.
![image](https://github.com/FarmGTEx/CattleCellAtlas_pipeline_V0/blob/main/Image/Figure%201.pdf)

2. Analysis pipeline

This repository contains analysis pipelines used by the CattleCA Consortium, including:

Raw sequencing data processing, quality control and cell clustering

Cell cycle analysis

Cell communication analysis

Gene regulatory network analysis

Pseudo time trajectory analysis

Cell type preference analysis

Sex bias analysis

High-dimensional weighted gene co-expression network analysis

Gene function enrichment analysis

Enrichment analysis between cell types and monogenic disorders

Enrichment analysis between cell types and complex traits

Linkage disequilibrium score regression analysis

3. Citation

Cattle Cell Atlas: a multi-tissue single cell expression repository for advanced bovine genomics and comparative biology (Preprint)
Bo Han1, #, Houcheng Li2, #, Qi Zhang1, #, Weijie Zheng1, #, Ao Chen1, #, Senlin Zhu3, #, Tao Shi4, #, Yali Hou5, #, Fei Wang6, #, Dong Zou7, Zhixiang Yuan7, Yu Song1, Wen Ye1, Aixia Du1, Yihan Fu1, Minghui Jia3, Zhonghao Bai2, Wansheng Liu8, Wenbin Tuo9, Jayne C. Hope10, David E. MacHugh11, 12, John Francis Oâ€™ Grady11, 12, Ole Madsen13, Goutam Sahana2, Yonglun Luo 14, 15, Lin Lin14, 15, Congjun Li16, Zexi Cai2, Lin Liu17, Zhang Zhang7, Zhu Ma17, George E. Liu16*, Huizeng Sun3*, Yu Jiang4*, Lingzhao Fang2*, Dongxiao Sun1* 

